pkg/crc/preflight/preflight_check_tray_darwin.go

```diff
@@ -0,0 +1,248 @@
+			<false/>
+		</dict>
+	</plist>`
+
+	daemonAgentLabel = "crc.daemon"
+	trayAgentLabel   = "crc.tray"
+)
+
+var (
+	launchAgentDir       = filepath.Join(constants.GetHomeDir(), "Library", "LaunchAgents")
+	daemonPlistFilePath  = filepath.Join(launchAgentDir, "crc.daemon.plist")
+	trayPlistFilePath    = filepath.Join(launchAgentDir, "crc.tray.plist")
+	stdOutFilePathDaemon = filepath.Join(constants.CrcBaseDir, ".crcd-agent.log")
+	stdOutFilePathTray   = filepath.Join(constants.CrcBaseDir, ".crct-agent.log")
+)
+
+type AgentConfig struct {
+	BinaryPath     string
+	StdOutFilePath string
+}
+
+func checkTrayExistsAndRunning() error {
+	logging.Debug("Checking if daemon plist file exists")
+	if !plistFileExists(daemonPlistFilePath) {
+		return errors.New("Daemon plist file does not exist")
+	}
+	logging.Debug("Checking if crc agent running")
+	if !agentRunning(daemonAgentLabel) {
+		return errors.New("crc daemon is not running")
+	}
+	logging.Debug("Checking if tray plist file exists")
+	if !plistFileExists(trayPlistFilePath) {
+		return errors.New("Tray plist file does not exist")
+	}
+	logging.Debug("Checking if tray agent running")
+	if !agentRunning(trayAgentLabel) {
+		return errors.New("Tray is not running")
+	}
+	return nil
+}
+
+func fixTrayExistsAndRunning() error {
+	// get the tray app
+	if !trayAppCached() {
+		err := downloadOrExtractTrayApp()
+		if err != nil {
+			return err
+		}
+	}
+	currentExecutablePath, err := goos.Executable()
+	if err != nil {
+		return err
+	}
+	daemonConfig := AgentConfig{
+		BinaryPath:     currentExecutablePath,
+		StdOutFilePath: stdOutFilePathDaemon,
+	}
+
+	trayConfig := AgentConfig{
+		BinaryPath:     filepath.Join(constants.CrcBinDir, constants.TrayBinaryName, "Contents", "MacOS", "CodeReady Containers"),
+		StdOutFilePath: stdOutFilePathTray,
+	}
+	logging.Debug("Creating daemon plist")
+	err = createPlist(daemonPlistTemplate, daemonConfig, daemonPlistFilePath)
+	if err != nil {
+		return err
+	}
+	logging.Debug("Creating tray plist")
+	err = createPlist(trayPlistTemplate, trayConfig, trayPlistFilePath)
+	if err != nil {
+		return err
+	}
+
+	// load crc daemon
+	err = loadPlist(daemonPlistFilePath)
+	if err != nil {
+		return err
+	}
+	if !agentRunning(daemonAgentLabel) {
+		if err = startAgent(daemonAgentLabel); err != nil {
+			return err
+		}
+	}
+	// load tray
+	err = loadPlist(trayPlistFilePath)
+	if err != nil {
+		return err
+	}
+	if !agentRunning(trayAgentLabel) {
+		if err = startAgent(trayAgentLabel); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func createPlist(template string, config AgentConfig, plistPath string) error {
+	var plistContent bytes.Buffer
+	t, err := goTemplate.New("plist").Parse(template)
+	if err != nil {
+		return err
+	}
+	err = t.Execute(&plistContent, config)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(plistPath, plistContent.Bytes(), 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func loadPlist(plistFilePath string) error {
```

Comment:
`launchPlist` may be a better function name?

Reply:
If this was `func (launch LaunchCtl) loadPlist(...)`, then we could name it `loadPlist`. In this case, it's just `loadPlist`, so we have no indication that `launchctl` is involved. With just `loadPlist`, I think the function reads the plist file in memory, maybe parsing it. I don't think it's going to pass it as an argument to `launchctl load`.